### PR TITLE
[Concurrency] Implement various improvements in the execution_state module [WiP]

### PR DESCRIPTION
**Pull request by lucasccordeiro**

### CODE REVIEW

**Comment by lucasccordeiro** [comment]
* This PR with `--incremental-bmc --no-por --context-switch 2 -no-goto-merge`: 

```
Statistics:            768 Files
  correct:             210
    correct true:       46
    correct false:     164
  incorrect:             1
    incorrect true:      0
    incorrect false:     1
  unknown:             557
  Score:               240 (max: 1171)

```
The only incorrect result: `goblint-regression/13-privatized_67-pthread_cond_wait_true`

* Master `--incremental-bmc --no-por --context-switch 2`:
```
Statistics:            768 Files
  correct:             420
    correct true:      137
    correct false:     283
  incorrect:            16
    incorrect true:      9
    incorrect false:     7
  unknown:             332
  Score:               157 (max: 1171)
```

**Comment by lucasccordeiro** [comment]
This PR with `--incremental-bmc --no-por --context-bound 2 --no-goto-merge`:

```
Statistics:            763 Files
  correct:             336
    correct true:       69
    correct false:     267
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:             427
  Score:               405 (max: 1160)
```

* Master `--incremental-bmc --no-por --context-switch 2`:
```
Statistics:            763 Files
  correct:             431
    correct true:      142
    correct false:     289
  incorrect:            15
    incorrect true:      9
    incorrect false:     6
  unknown:             317
  Score:               189 (max: 1160)
```

* Master `--incremental-bmc --no-por --context-switch 2`:
```
Statistics:            768 Files
  correct:             420
    correct true:      137
    correct false:     283
  incorrect:            16
    incorrect true:      9
    incorrect false:     7
  unknown:             332
  Score:               157 (max: 1171)
```

File: src/esbmc/options.cpp [hunk 1]
```diff
@@ -245,6 +245,10 @@ const struct group_opt_templ all_cmd_options[] = {
      boost::program_options::value<int>()->default_value(-1)->value_name("nr"),
      "limit number of context switches for each thread"},
     {"state-hashing", NULL, "enable state-hashing, prunes duplicate states"},
+    {"no-goto-merge",
```
Comment (kunjsong01) [inline src/esbmc/options.cpp:248]: would have been nice to have a test case that use this new option.

Comment (lucasccordeiro) [reply src/esbmc/options.cpp:248]: Done.

**Pull request merged**